lib/posix/pthread.c

```diff
@@ -254,31 +407,32 @@ int pthread_setcancelstate(int state, int *oldstate)
  */
 int pthread_cancel(pthread_t pthread)
 {
-	struct posix_thread *thread = to_posix_thread(pthread);
+	void *retval;
+	int detachstate;
 	int cancel_state;
-	k_spinlock_key_t cancel_key;
+	k_spinlock_key_t key;
+	struct posix_thread *t;
 
-	if ((thread == NULL) || (thread->state == PTHREAD_TERMINATED)) {
+	t = to_posix_thread(pthread);
+	if (t == NULL) {
 		return ESRCH;
 	}
 
-	cancel_key = k_spin_lock(&thread->cancel_lock);
-	thread->cancel_pending = 1;
-	cancel_state = thread->cancel_state;
-	k_spin_unlock(&thread->cancel_lock, cancel_key);
-
+	key = k_spin_lock(&q_lock);
+	t->cancel_pending = true;
+	cancel_state = t->cancel_state;
+	detachstate = t->detachstate;
 	if (cancel_state == PTHREAD_CANCEL_ENABLE) {
-		pthread_mutex_lock(&thread->state_lock);
-		if (thread->state == PTHREAD_DETACHED) {
-			thread->state = PTHREAD_TERMINATED;
-		} else {
-			thread->retval = PTHREAD_CANCELED;
-			thread->state = PTHREAD_EXITED;
-			pthread_cond_broadcast(&thread->state_cond);
+		if (detachstate == PTHREAD_CREATE_DETACHED) {
+			t->retval = PTHREAD_CANCELED;
 		}
-		pthread_mutex_unlock(&thread->state_lock);
+	}
+	retval = t->retval;
+	k_spin_unlock(&q_lock, key);
 
-		k_thread_abort(&thread->thread);
+	if (cancel_state == PTHREAD_CANCEL_ENABLE) {
+		posix_thread_finalize(t, retval);
```

Comment:
This appears to assume that `pthread_cancel()` is only called on `pthread_self()`, which I don't believe is true. I think that means `posix_thread_finalize()` shouldn't call `k_thread_abort()` and that `pthread_exit()` should do that instead, after calling `posix_thread_finalize()`.

Reply:
Will check it out.

Reply:
Yes - good catch - fixed!